src/Juvix/Compiler/Core/Translation/FromSource.hs

```diff
@@ -538,17 +651,42 @@ parseLocalName = parseWildcardName <|> parseIdentName
       (txt, i) <- identifierL
       lift $ freshName KNameLocal txt i
 
+exprPi ::
+  Members '[Reader ParserParams, InfoTableBuilder, NameIdGen] r =>
+  Index ->
+  HashMap Text Index ->
+  ParsecS r Node
+exprPi varsNum vars = do
+  kw kwPi
+  name <- parseLocalName
+  kw kwColon
+  ty <- expr varsNum vars
+  kw kwComma
+  let vars' = HashMap.insert (name ^. nameText) varsNum vars
+  body <- expr (varsNum + 1) vars'
+  return $ mkPi (binderInfo name ty) ty body
+
 exprLambda ::
   Members '[Reader ParserParams, InfoTableBuilder, NameIdGen] r =>
   Index ->
   HashMap Text Index ->
   ParsecS r Node
 exprLambda varsNum vars = do
   lambda
-  name <- parseLocalName
+  (name, mty) <- lambdaName
   let vars' = HashMap.insert (name ^. nameText) varsNum vars
-  body <- expr (varsNum + 1) vars'
-  return $ mkLambda (binderNameInfo name) body
+  body <- bracedExpr (varsNum + 1) vars'
+  return $ mkLambda (binderInfo name (fromMaybe mkDynamic' mty)) body
+  where
+    lambdaName =
+      parens
+        ( do
+            n <- parseLocalName
+            kw kwColon
+            ty <- expr varsNum vars
+            return (n, Just ty)
+        )
+        <|> (parseLocalName >>= \n -> return (n, Nothing))
```

Comment:
`(\n -> (n, Nothing)) <$> parseLocalName`

Reply:
Okay, I'll change it. Though I never quite got the point of trying so hard to be so very terse and "point-free" and use all those funny operator symbols. It seems that's what HLint and experienced Haskell programmers recommend. But it makes Haskell look like Perl and hard to read sometimes.

Reply:
I would say `<$>` is common in the haskell ecosystem